kmm/positions/wire_camera_positions.py

```diff
@@ -1,47 +1,54 @@
+import numpy as np
 from pydantic import validate_arguments
 
 from kmm import CarDirection
 from kmm.positions.positions import Positions
 
 
 @validate_arguments(config=dict(arbitrary_types_allowed=True))
-def wire_camera_positions(positions: Positions, direction: CarDirection):
-    ascending = kmm_ascending(positions.dataframe)
-
-    if (
-        (direction == CarDirection.A and ascending)
-        or (direction == CarDirection.B and not ascending)
-    ):
-        correction = -8
-
-    elif (
-        (direction == CarDirection.A and not ascending)
-        or (direction == CarDirection.B and ascending)
-    ):
-        correction = 8
+def wire_camera_positions(positions: Positions, car_direction: CarDirection):
 
+    if car_direction == CarDirection.A:
+        corrections = kmm_directions(positions.dataframe) * -8
+    elif car_direction == CarDirection.B:
+        corrections = kmm_directions(positions.dataframe) * 8
     else:
-        raise ValueError
+        raise ValueError(car_direction)
 
     return positions.replace(
         dataframe=(
             positions.dataframe
-            .assign(meter=lambda df: df["meter"] + correction)
+            .assign(meter=lambda df: df["meter"] + corrections)
         )
     )
 
 
-def kmm_ascending(dataframe):
-    total_meter = dataframe["kilometer"] * 1000 + dataframe["meter"]
-    diff = total_meter.values[:-1] - total_meter.values[1:]
-    descending = (diff < 0).mean()
-    ascending = (diff > 0).mean()
+def kmm_directions(df):
 
-    if descending < 0.9 and ascending < 0.9:
-        raise ValueError("Unable to determine ascending/descending kmm numbers")
+    total_meter = (df["kilometer"] * 1000 + df["meter"]).values
 
-    else:
-        return ascending > 0.9
+    track_section_changes = np.argwhere(np.concatenate([
+        np.array([True]),
+        df["track_section"].values[1:] != df["track_section"].values[:-1],
+    ])).squeeze(1).tolist() + [len(df)]
+
+    return np.concatenate([
+        (
+            np.ones(to_index - from_index, dtype=np.uint8)
+            * kmm_direction(total_meter[from_index: to_index])
```

Comment:
Do we need to do this in vanilla python?

Reply:
> Is it possible to do something like:
> 
> ```
> np.sign(positions.dataframe["meter"].iloc[:-1] - positions.dataframe["meter"].iloc[1:])
> ```
> 
> Or group by "track_section" and "kilometer" and find min + max with their respective "centimeters"?
> 
> ```
> positions.dataframe.groupby(["track_section", "kilometer"])["meter"].min()
> positions.dataframe.groupby(["track_section", "kilometer"])["meter"].max()
> ```

Ah I like the groupby idea :) It would be harder to detect corrupt kmm's though like I do now with
```
if len(diffs) >= 10 and (diffs > 0).mean() < 0.9 and (diffs < 0).mean() < 0.9:
    raise ValueError("Unable to determine direction of kmm numbers.", diffs)
```
but maybe we don't have to care about that. If so I would actually just group by `track_section` and take the difference between `idxmax` of `total_meter` as ascending/descending. Maybe then we would want to accomodate for >1000m kms though.

Reply:
It also assumes that the train only goes over a track section once per measurement though, do we know that for sure? I don't know exactly how a track section is defined, for example I could imagine this being a realistic scenario
```
               ______ track_section_B
             /        \
-------------------------------track_section_A
```